x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,348 @@
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// handle message
+func handleMessage(m sqs.Message, bucketNames []string) (s3Infos []s3Info, err error) {
+	msg := map[string]interface{}{}
+	err = json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		err = errors.Wrap(err, "json unmarshal sqs message body failed")
+		return
+	}
+
+	records := msg["Records"].([]interface{})
+	for _, record := range records {
+		recordMap := record.(map[string]interface{})
+		if recordMap["eventSource"] == "aws:s3" && recordMap["eventName"] == "ObjectCreated:Put" {
+			s3Info := s3Info{}
+			if !stringInSlice(recordMap["awsRegion"].(string), bucketNames) {
+				continue
+			}
+
+			s3Info.region = recordMap["awsRegion"].(string)
+			s3Record := recordMap["s3"].(map[string]interface{})
+
+			bucketInfo := s3Record["bucket"].(map[string]interface{})
+			s3Info.name = bucketInfo["name"].(string)
+
+			objectInfo := s3Record["object"].(map[string]interface{})
+			s3Info.key = objectInfo["key"].(string)
+			s3Infos = append(s3Infos, s3Info)
+		}
+	}
+	return
+}
+
+// stringInSlice checks if a string is already exists in list
+// If there is no bucketNames configured, then collect all.
+func stringInSlice(name string, bucketNames []string) bool {
+	if bucketNames == nil || len(bucketNames) == 0 {
+		return true
+	}
+
+	for _, v := range bucketNames {
+		if v == name {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *Input) readS3Object(svc s3iface.S3API, s3Infos []s3Info) ([]*beat.Event, error) {
+	var events []*beat.Event
+	if len(s3Infos) > 0 {
+		var wg sync.WaitGroup
+		numS3Infos := len(s3Infos)
+		wg.Add(numS3Infos)
+
+		for i := range s3Infos {
+			go func(s3Info s3Info) {
+				// launch goroutine to handle each message
+				defer wg.Done()
+
+				s3GetObjectInput := &s3.GetObjectInput{
+					Bucket: awssdk.String(s3Info.name),
+					Key:    awssdk.String(s3Info.key),
+				}
+				req := svc.GetObjectRequest(s3GetObjectInput)
+
+				resp, err := req.Send()
+				if err != nil {
+					p.logger.Error(errors.Wrap(err, "s3 get object request failed"))
+				}
+
+				reader := bufio.NewReader(resp.Body)
+				line := 0
+				for {
+					log, err := reader.ReadString('\n')
+					if err != nil {
+						if err == io.EOF {
+							line++
+							event := createEvent(log, int64(line), s3Info)
+							events = append(events, event)
+							break
+						} else {
+							p.logger.Error(errors.Wrap(err, "ReadString failed"))
+						}
+					}
+					// create event per log line
+					line++
+					event := createEvent(log, int64(line), s3Info)
+					events = append(events, event)
```

Comment:
great change making use of a streaming reader! there is something else needed to avoid high memory consumption: You are streaming the lines as you read them, but then you are putting storing them on an array and processing all of them at the same time, before pushing them to the output. That means that this code is still storing all events in RAM before sending them to the output.

With some more code reorganization you could push each event as soon as it's read from S3, avoiding extra unneeded memory consumption (and time)

Reply:
Yeah good point, with the current code, I'm still collecting all events from one log file before pushing it, which is a waste for the streaming reader purpose... I will make the change! Thank you!!